### ISSUE TITLE
Analyzer for XML comments?

### ISSUE THREAD

**Issue by hansmbakker** [dotnet/roslyn#50694]
**Brief description:**

Analyzer that checks whether
1. all public methods have been documented, or inherit a xml comment using `<inheritDoc />`.
    - if class inherits an interface/class, suggestion to use `<inheritDoc />` if it does not use it yet
1. the method signature (return + parameters) matches the xml comments
1. the `param name`s of the xml comment match the parameters of the method

**Languages applicable:**
I believe XML Comments are a C# thing?

**Code example that the analyzer should report:**

**Case 1**

Bad code (public method missing documentation)
```csharp
    /// <summary>
    /// Simple interface to test with
    /// </summary>
    public interface ITestInterface
    {
        void Test();
    }
```

Bad code (public type missing documentation)
```csharp
    public interface ITestInterface
    {
        /// <summary>
        /// Simple method to test with
        /// </summary>
        void Test();
    }
```

Bad code (public class and method are missing documentation, which can be fixed by using `<inheritDoc cref="ITestInterface" />`
```csharp
    /// <summary>
    /// Simple interface to test with
    /// </summary>
    public interface ITestInterface
    {
        /// <summary>
        /// Simple method to test with
        /// </summary>
        void Test();
    }

    public class TestClass : ITestInterface
    {
        public void Test() { }
    }
```

**Case 2**

Bad code (`returns` shouldn't be here)
```csharp
        /// <summary>
        /// Comment
        /// </summary>
        /// <returns></returns>
        public void Test()
```

Bad code (`returns` missing)
```csharp
        /// <summary>
        /// Comment
        /// </summary>
        public int Test()
```

Bad code (wrong amount of parameters documented)
```csharp
        /// <summary>
        /// Comment
        /// </summary>
        /// <param name="testParam1"></param>
        public void Test(int testParam1, string testString)
```

**Case 3**
Bad code (parameter name does not match)
```csharp
        /// <summary>
        /// Comment
        /// </summary>
        /// <param name="testParam2"></param>
        public void Test(int testParam1, string testString)
```

**Documentation requirements:**

When this analyzer is implemented, it must be documented by following the steps at [Documentation for IDE CodeStyle analyzers](https://github.com/dotnet/roslyn/blob/master/docs/contributing/Documentation%20for%20IDE%20CodeStyle%20analyzers.md).

**Comment by Youssef1313** [comment]
There is already a compile warning for undocumented public APIs.

https://docs.microsoft.com/dotnet/csharp/language-reference/compiler-messages/cs1591

**Comment by hansmbakker** [comment]
oh, you're right! I didn't know about that.

I see these warnings only come up when you specifically configure an XML documentation file in the project though... it would be nice if that would be part of the .Net analyzers though, so you could configure it in an intuitive way

**Comment by sharwell** [comment]
> I see these warnings only come up when you specifically configure an XML documentation file in the project though

When XML documentation file is disabled, the compiler treats `///` comments the same way as `//` comments (i.e. there are no documentation comments in a project unless the project enables the XML documentation file).

**Comment by hansmbakker** [comment]
So this is "as designed", then.. I find it not very intuitive (since Intellisense needs the XML format rather than normal comments so there is value in the XML comments even when the XML file is disabled), but then I guess that changing this will be a high barrier if possible at all.
